File: algorithms/problem-solving/target-sum/target-sum.go
package main

import (
	"fmt"
	"sort"
)

func twoNumberSum1(array []int, target int) []int {

	xs := []int{}
	for i := 0; i < len(array)-1; i++ {
		firstNum := array[i]
		for j := i + 1; j < len(array); j++ {
			secondSum := array[j]
			if firstNum+secondSum == target {
				xs = append(xs, firstNum, secondSum)
			}

		}
	}
	return xs
}

func twoNumberSum2(array []int, target int) []int {
	xs := []int{}

	nums := make(map[int]bool)

	for _, n := range array {
		r := target - n
		_, ok := nums[r]
		if ok == true {
			xs = append(xs, r, n)
		} else {
			nums[n] = true
		}
	}
	return xs
}
func twoNumberSum3(array []int, target int) []int {
	sort.Ints(array)

	left, right := 0, len(array)-1

	for left < right {
		current := array[left] + array[right]
		if target == current {
			return []int{array[left], array[right]}
		} else if current < target {
			left++
		} else if current > target {
			right--
		}
	}

	return []int{}
}

func main() {
	xs := []int{3, 5, -4, 8, 11, 1, -1, 6}
	res := twoNumberSum3(xs, 10)
	fmt.Println(res)
}

File: algorithms/problem-solving/target-sum/target-sum.ts
// O(n^2) Speed | O(1) Space
export function twoNumberSum1(array: number[], targetSum: number) {
  for (let i = 0; i < array.length - 1; i++) {
    const firstNumber = array[i]
    console.log(firstNumber)
    for (let j = i + 1; j < array.length; j++) {
      const secondNumber = array[j]
      if (firstNumber + secondNumber === targetSum) {
        return [firstNumber, secondNumber]
      }
    }
  }
}

// O(n) Speed | O(n) Space
export function twoNumberSum2(array: number[], targetSum: number) {
  const nums: Record<string, boolean> = {}
  for (const n of array) {
    const sum = targetSum - n
    if (nums[sum]) {
      return [sum, n]
    } else {
      nums[n] = true
    }
  }
  return []
}

// O(nLog(n)) Speed | O(1) Space
export function twoNumberSum3(array: number[], targetSum: number) {
  const xs = array.sort((a, b) => a - b)
  let left = 0
  let right = array.length - 1

  while (left < right) {
    const currentSum = xs[left] + xs[right]
    if (currentSum === targetSum) {
      return [xs[left], xs[right]]
    } else if (currentSum < targetSum) {
      left++
    } else if (currentSum > targetSum) {
      right--
    }
  }
  return []
}

// console.log(twoNumberSum1([3, 5, -4, 2], 7))
console.log(twoNumberSum3([3, 5, -4, 8, 11, 1, -1, 6], 10))
// console.log(twoNumberSum2([9, 12, -5, 2], 4))
